designs/spot-consolidation.md

```diff
@@ -0,0 +1,112 @@
+
+Karpenter’s consolidation algorithm minimizes the price of EC2 instances given pod scheduling constraints. However, price minimization presents challenges for EC2 Spot, which offers a tradeoff between price and interruption rate. Karpenter uses EC2 Spot’s PriceCapacityOptimized (PCO) allocation strategy to provide a balanced approach to this tradeoff, and doesn't always select the cheapest instance type, since it may result in immediate interruption. If Karpenter were to consolidate spot capacity based purely on cost, it would repeatedly interrupt the same instance, walking down the PCO decision ladder, until only the lowest price remained. Because of these complications, Karpenter does not currently support "spot replacement" consolidation.
+
+All consolidation actions contain an inherent tradeoff between price and availability. Consolidation actions are equivalent to a spot interruption, but without the deadline. This tradeoff applies to on-demand capacity, as well. This document explores approaches to enable consolidation for spot instances without regressing to lowest price. 
+
+Each of the below solution works on a certain heuristics because most cloud providers don't surface capacity pool information for spot which presents unique challenges to efficiently select a certain spot instance Type that would cater to better price and availability with respect to the current node. Hence, the idea is to enable the consolidation based on different heuristics so that it makes sense to re-size the nodes only when there is significant cost benefit. 
+
+## Design Options
+
+### 1: Minimum Flexibility [Recommended] 
+
+Karpenter will replace a spot node with a cheaper spot node for both single and multi-node consolidation but with an additional condition for single-node consolidation. Cheaper spot instance types are selected with the `price-capacity-optimized` strategy and often the cheapest spot instance type is not launched due to the likelihood of interruption. Karpenter would replace spot node for single-node consolidation only if there are more than 15 potential replacement spot instance types. Karpenter would have to limit its flexibility to this threshold to avoid repeat interruptions, e.g. "walking down the ladder". By ensuring that we have a minimum flexibility in our launch decisions, we can be assured that we have enough instance diversity that it's very unlikely that all instances that we select will have high interruption. Further, Karpenter would have to limit its flexibility during the launch scenario as well, else face an immediate consolidation if the launched instance type’s price were more than the price of any instance type under the flexibility threshold. Since on-demand capacity uses the lowest price strategy, its behavior is unaffected if these rules are applied equally to both spot and on-demand launches.
```

Comment:
It might be nice to callout why we made this decision _only_ for single node consolidation. Basically, the issue here is that we basically think that multi-node consolidation decisions that required the minimum instance type flexibility wouldn't occur frequently enough if we enforced it. On top of this, there's some considerations that we should take into account with our current handling for mutli-node spot such as the mixed capacity type scenario that it might be good to callout here. 

In either case, I'd expand on the rationale with why we went this direction so that it's a bit more fleshed out.

Reply:
Added more details regarding why we chose this flexibility only for single-node and why did we decide to do it only in the consolidation path.